### PR TITLE
fix: expose manipulate pipes command

### PR DESCRIPTION
**Pull request by polvalente**
This PR also adds a test that shows a bug which seems to be in Macro.to_string itself :( I've opened an issue in the Elixir repo https://github.com/elixir-lang/elixir/issues/10864 for discussion. Maybe this is expected and is something we have to deal with

### CODE REVIEW

**Comment by polvalente** [comment]
It seems that this fix is needed: https://github.com/elixir-lang/elixir/issues/10864#issuecomment-812690698
Let's wait for a response in the issue before merging this PR

**Comment by polvalente** [comment]
So, a commit was added changing this behavior in Elixir. Due to the fact that we need to be retrocompatible, I've added a function to treat added escapes whenever they are added. The main assumption is that `from_pipe` and `to_pipe` should never change the number of non-whitespace chars besides adding or removing `|>`, so if the number of `\` changes, it means that we need to execute said replacement.

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex
Comment (polvalente) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex:68]: This fix doesn't work properly because sigil delimiters changed to `"` prior to 1.10

Some parsing of the AST will be needed to treat sigils

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex
Comment (polvalente) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex:68]: This is an attempt to handle escape characters as expected. Local tests passes successfully, but I'm not sure if this is the best way to approach the problem

**Comment by axelson** [comment]
I've been trying this out a bit with (https://github.com/elixir-lsp/vscode-elixir-ls/pull/182/files) and I'm running into some issues. It appears that sometimes the extracted text . Here's a partial failing test (partial because it should check the result):

```elixir
    test "can handle multiple calls" do
      {:ok, _} =
        JsonRpcMock.start_link(success_reply: {:ok, %{"applied" => true}}, test_pid: self())

      uri = "file:/some_file.ex"

      text = """
      defmodule BasicEx do
        def add(a) do
          require Logger

          Logger.info(to_string(add_num(a, 12)))
        end

        def add_num(a, num), do: a + num
      end
      """

      assert {:ok, nil} =
               ManipulatePipes.execute(
                 ["fromPipe", uri, 4, 16],
                 %Server{
                   source_files: %{
                     uri => %SourceFile{
                       text: text
                     }
                   }
                 }
               )
```

That test results in:
```
code_string: "Logger.info(to_string(add_num(a, 12))"

  1) test execute/2 fromPipe can handle multiple calls (ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipesTest)
     test/providers/execute_command/manipulate_pipes_test.exs:763
     ** (TokenMissingError) nofile:1: missing terminator: ) (for "(" starting at line 1)
     code: ManipulatePipes.execute(
     stacktrace:
       lib/language_server/providers/execute_command/manipulate_pipes/ast.ex:23: ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipes.AST.from_pipe/1
       lib/language_server/providers/execute_command/manipulate_pipes.ex:122: ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipes.from_pipe_at_cursor/3
       lib/language_server/providers/execute_command/manipulate_pipes.ex:31: ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipes.execute/2
       test/providers/execute_command/manipulate_pipes_test.exs:782: (test)
```
Where I would expect the value of `code_string` to instead be more like `Logger.info(to_string(add_num(a, 12)))` (in actuality it is missing a paren). I think a good test would be to run the same command, but once with each character position and ensure that the same value is output each time (which is what I would expect as a user running `extension.toPipe` on a line without a `;`)

**Comment by polvalente** [comment]
@axelson could you elaborate on "once with each character position"? Do you mean something like `for column <- 0..line_length-1, do: ...`, or just the "valid" columns for a given "to pipe"/"from pipe"?

The command is currently highly dependent on where the cursor is because there can be multiple function calls and multple pipes in a line. Anyway, your case is indeed a bug and I'll investigate on what's causing it

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex
Comment (polvalente) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:241]: @axelson I've added this because your test actually was calling "fromPipe", which shouldn't do anything if there aren't any pipes in the selected text

Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:241]: Another possible option which would require some rework is to scan until a pipe is found, but this could have unexpected results

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex
Comment (polvalente) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:295]: This was needed if the suggested test is changed to "toPipe", because `Logger.info(to_string(...)` doesn't have any spaces, but the first `(` actually breaks the identifier `to_string`

File: apps/language_server/test/providers/execute_command/manipulate_pipes_test.exs
Comment (polvalente) [inline apps/language_server/test/providers/execute_command/manipulate_pipes_test.exs:434]: This one might bring up some discussion. The way that our macro-walking function is implemented, the code infers that the innermost function call is the one that should be piped.

Does this need to be reworked so the outermost one is the piped call?

**Comment by axelson** [comment]
> @axelson could you elaborate on "once with each character position"? Do you mean something like for column <- 0..line_length-1, do: ..., or just the "valid" columns for a given "to pipe"/"from pipe"?

Yeah, `for column <- 0..line_length-1, do: ...` is exactly what I mean. But since the output may change based on where the cursor is, it would just check that no exception is raised.

> The command is currently highly dependent on where the cursor is because there can be multiple function calls and multple pipes in a line. Anyway, your case is indeed a bug and I'll investigate on what's causing it

Yeah, makes sense :+1:

File: apps/language_server/test/providers/execute_command/manipulate_pipes_test.exs
Comment (polvalente) [inline apps/language_server/test/providers/execute_command/manipulate_pipes_test.exs:893]: @axelson I've added this helper assertion that travels through the whole text calling the corresponding command ("toPipe" or "fromPipe"). This should ensure that no position in each test file raises in the code. Some bugs have been found and fixed this way already :)

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex
Comment (lukaszsamson) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex:87]: add a TODO with info which part may be removed when we require elixir 1.13 (or 1.12 if the fix gets backported)

File: apps/language_server/test/providers/execute_command/manipulate_pipes_test.exs
Comment (lukaszsamson) [reply apps/language_server/test/providers/execute_command/manipulate_pipes_test.exs:446]: I guess that would be preferred. On the other hand this is an ambiguous case.

Comment (polvalente) [reply apps/language_server/test/providers/execute_command/manipulate_pipes_test.exs]: @lukaszsamson I've fixed this in the latest commit. The issue was that there was a `and t != []` clause which disallowed 1-arg functions from being piped. No other tests broke, so that's a good sign, I guess

**Review by axelson — CHANGES_REQUESTED** [review]
Okay, taking a deeper look at this. Sorry for the delay.

I have a high-level comment, that I wish I had flagged earlier. Given this snippet:

``` elixir
Ash.Changeset.for_create(Track, :create, data)
|> GenTracker.API.create()
```

Running `toPipe` on the first line I would expect to end up with

``` elixir
Track
|> Ash.Changeset.for_create(:create, data)
|> GenTracker.API.create()
```

But instead end up with

``` elixir
Track |> Ash.Changeset.for_create(:create, data)
|> GenTracker.API.create()
```

But I suppose this would be hard to handle for code that has multiple function calls on one line, so perhaps it's best handled as a future improvement.

Also (with the above code) I would expect the same result no matter where on the first line I called `toPipe`, but instead there's a bug where if you call `toPipe` at the end of the line you end up with:

``` elixir
Ash.Changeset.for_create(Track, :create, data)GenTracker.API.create()
```

Which is obviously wrong and does not compile.

`to_pipe_at_cursor` and `from_pipe_at_cursor` suggested changes:
- For easier testing I think we should make them public functions and call them from the tests.
- I think we should have those functions return a `%ElixirLS.LanguageServer.Protocol.TextEdit{}` (which needs to be created) instead a raw map
- Then those two can be combined with `ElixirLS.LanguageServer.Test.TestUtils.apply_text_edit/2` to greatly simplify the tests
- I have some code related to this that I hope to post soon

**Comment by polvalente** [comment]
> Okay, taking a deeper look at this. Sorry for the delay.
> 
> I have a high-level comment, that I wish I had flagged earlier. Given this snippet:
> 
> ```elixir
> Ash.Changeset.for_create(Track, :create, data)
> |> GenTracker.API.create()
> ```
> 
> Running `toPipe` on the first line I would expect to end up with
> 
> ```elixir
> Track
> |> Ash.Changeset.for_create(:create, data)
> |> GenTracker.API.create()
> ```
> 
> But instead end up with
> 
> ```elixir
> Track |> Ash.Changeset.for_create(:create, data)
> |> GenTracker.API.create()
> ```
> 
> But I suppose this would be hard to handle for code that has multiple function calls on one line, so perhaps it's best handled as a future improvement.

Thanks for the review!

About changing multi-line to single-line pipes, I'd rather leave that as a future improvement, since most [citation needed] people that will be using this command will also be using "format on save" or similiar features.

> Also (with the above code) I would expect the same result no matter where on the first line I called `toPipe`, but instead there's a bug where if you call `toPipe` at the end of the line you end up with:
> 
> ```elixir
> Ash.Changeset.for_create(Track, :create, data)GenTracker.API.create()
> ```
> 
> Which is obviously wrong and does not compile.

Nice catch! I'll add a test and a fix for this as soon as I can get back to this PR.
 
> `to_pipe_at_cursor` and `from_pipe_at_cursor` suggested changes:
> 
> * For easier testing I think we should make them public functions and call them from the tests.

Which tests are you talking about? The integration ones, which receive the command itself?
I guess we could migrate them to a model where there is one test-case which calls it and the others only call these functions. WDYT?

> * I think we should have those functions return a `%ElixirLS.LanguageServer.Protocol.TextEdit{}` (which needs to be created) instead a raw map
> * Then those two can be combined with `ElixirLS.LanguageServer.Test.TestUtils.apply_text_edit/2` to greatly simplify the tests

Sounds good!

> * I have some code related to this that I hope to post soon

Should I wait for said code?

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex
Comment (polvalente) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:152]: @axelson here's what caused the tests you added to break: do to how `do_get_function_call` works, when we're at a closing paren, it will jump and try to pipe the next function call. The same for when we're already at the end of a line.

I'm thinking that maybe the correct fix is to track back the text until we reach a "(" character instead (updating indices properly).

I haven't tested, but this one probably breaks if the line ends as ") \n"

WDYT? Any other ideas?

Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:152]: Actually, tracing back blind yields other problems when the cursor is in the leading whitespace for the line :(

Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:152]: I"ve added a test which demonstrates the ")\s\n" case still breaking

Comment (axelson) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:152]: > do to how do_get_function_call works, when we're at a closing paren, it will jump and try to pipe the next function call. The same for when we're already at the end of a line.

Do we need to keep the "jumping forward" behavior? What is it buying us?

Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:152]: It's more of an artifact on how the "function call finding" code works. for instance on a multi-line function call like:
```
    f(
    x,
    y,
    z)
```

Having the cursor at the beggining of the line containing `x` will still find the function correctly.

However, we can just for the cursor to be, for instance at a "(" character or a non-whitespace character, for instance

Comment (axelson) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:152]: ah, I see.

FYI, I found another breaking case so I'm working on fixing that.

**Comment by axelson** [comment]
@lukaszsamson Would you mind taking another look at the code as well? It's changed a bit since your review.

**Review by axelson — APPROVED** [review]

**Comment by lukaszsamson** [comment]
@axelson Looks good to me, LGTM

**Pull request merged**
